core/repo_status.py

```diff
@@ -0,0 +1,249 @@
+import threading
+import uuid
+
+
+import sublime
+from sublime_plugin import TextCommand, EventListener
+
+from .git_command import GitCommand, repo_path_for_view
+from .git_mixins.status import FileStatus, MERGE_CONFLICT_PORCELAIN_STATUSES
+from . import state
+
+
+if False:
+    from typing import Callable, Dict, Iterable, List, Optional, TypeVar
+    from .state import RepoPath, RepoStatus
+
+    T = TypeVar('T')
+    Thunk = Callable[[], None]
+
+
+_lock = threading.Lock()
+filter_ = partial(filter, None)  # type: (Callable[[Iterable[Optional[T]]], Iterable[T]])
+
+active_view = None  # type: Optional[sublime.View]
+current_token = {}  # type: Dict[RepoPath, str]
+
+
+class GsRegularStatusUpdaterEventListener(EventListener):
+    def on_new_async(self, view):
+        maybe_update_status_for_view(view)
+
+    # Sublime calls `on_post_save_async` events only on the primary view.
+    # We thus track the state of the `active_view` manually so that we can
+    # refresh the status bar of cloned views.
+
+    # Note: We listen for 'on_activated' bc we must draw on the main
+    # thread for the happy path to avoid visual yank on the status bar.
+    # We also only draw if the repo path is cached bc we must avoid
+    # doing expensive work on the main thread.
+    def on_activated_async(self, view):
+        global active_view
+        active_view = view
+
+        maybe_update_status_for_view(view)
+
+    def on_post_save_async(self, view):
+        global active_view
+        if active_view and active_view.buffer_id() == view.buffer_id():
+            maybe_update_status_for_view(active_view)
+        else:
+            maybe_update_status_for_view(view)
+
+
+class gs_update_status_bar(TextCommand):
+    def run(self, edit):
+        sublime.set_timeout_async(partial(maybe_update_status_for_view, self.view))
+
+
+def maybe_update_status_for_view(view):
+    """Record intent to update the status bar."""
+    if view_is_transient(view):
+        return
+
+    repo_path = repo_path_for_view(view)
+    if repo_path:
+        maybe_update_status_async(repo_path)
+
+
+def view_is_transient(view: sublime.View) -> bool:
+    """Return whether a view can be considered 'transient'.
+
+    For our purpose, transient views are 'detached' views or widgets
+    (aka quick or input panels).
+    """
+
+    # 'Detached' (already closed) views don't have a window.
+    window = view.window()
+    if not window:
+        return True
+
+    # Widgets are normal views but the typical getters don't list them.
+    group, index = window.get_view_index(view)
+    if group == -1:
+        return True
+
+    return False
+
+
+def maybe_update_status_async(repo_path):
+    # type: (RepoPath) -> None
+    with _lock:
+        current_token[repo_path] = token = uuid.uuid4().hex
+    sink = partial(update_status, repo_path)
+    pred = lambda: current_token.get(repo_path) == token
+    sublime.set_timeout_async(
+        partial(executor, pred, sink)
+    )
```

Comment:
The implementation of `pred` and `executor` have changed a lot during this PR but the structure to have a `pred`, the `executor`, and the `sink` actually never changed. So I wouldn't do this. (Generally lambdas in python are hard to read except for the trivial cases. I usually avoid them because they're just not flexible enough to change.)

Reply:
GitSavvy code is all full of lambdas on calls to `sublime.set_timeout_async` and similar functions that accept a callable.

Personally, lambda version reads fine to me, but of course defining a function that does the same is fine as well.

The executor, however, reads badly. One needs to follow the code a few times back and forth, at least the first time (just in case executor does something unexpected), as well as be familiar (or spend time looking up) some new concepts or abstractions (sink? thunk? pred? Thunk?) ... for such a short piece of code it seems like it's a shame to introduce all of these.

I think this sort of addition should be discussed (e.g with other maintainers) last we introduce too many paradigms for the project... mixed methods of code flow are just as confusing. It may well be that refactoring the whole project to use this sort of alternative "functional flow control" will result in much cleaner code, but then we should probably do that in a separate PR.

Reply:
You're correct that in the current state executor is 'refactored to an end', where it becomes ridiculous, strange or funny. This is still for changeability. Maybe we need a lock here. Everytime you hack on it, I add print messages to it e.g. to catch the `else: print('--> skipped updating'). 

So, *yes* the executor code became so short that I would write it probably different. *no* this is at the beginning of the lifecycle for the global app state. We should focus on exploring and changing.